### PR TITLE
[CLI] fix a few things in domain migration command

### PR DESCRIPTION
**Pull request by shijiesheng**
**What changed?**

- add two methods in ClientFactory for migration
- countWorkflowCheck should be on current domain
- fix concurrency bug
- make template look slightly better
- refactor to move out domainMigration into separate cli entity

**Why?**

`add two methods in ClientFactory for migration`
This is needed to be able to use environment flags for internal usage

**How did you test it?**
```
shengs@shengs-C02XN3VDJGH6 cadence % go run cmd/tools/cli/main.go --do domain-1 domain migration --destination_address localhost:7933 --destination_domain domain-2 --tasklist ts1 --search_attr someID:INT
2023/08/08 10:21:28 Loading configFiles=[config/base.yaml config/development.yaml]
2023/08/08 10:21:28 Loading configFiles=[config/base.yaml config/development.yaml]
{"level":"info","ts":"2023-08-08T10:21:28.241-0700","msg":"Updated dynamic config","logging-call-at":"file_based_client.go:284"}
Validation Check:
- Domain Meta Data: true
    Current Domain:
      Name: domain-1
      UUID: d3cc0a3c-8f55-40af-9820-c48f8a0f46e5
    New Domain:
      Name: domain-2
      UUID: 203f0088-1d0f-40ac-a7c3-fdcf57ec0a99
- Workflow Check: true
    Long Running Workflow Num: 0
- Dynamic Config Check: false
    - Config Key: matching.asyncTaskDispatchTimeout
        Current Response:
          Data: "50s"
          Filters:
            - Name: domainName
              Value: "domain-1"
            - Name: taskListName
              Value: ["ts1"]
        New Response:
          Data: "20s"
          Filters:
            - Name: domainName
              Value: "domain-2"
            - Name: taskListName
              Value: ["ts1"]
- Search Attributes Check: false
    Missing Search Attributes in Current Domain:
      - someID:INT
    Missing Search Attributes in New Domain:
      - someID:INT
```

**Potential risks**

No Risk

**Release notes**

**Documentation Changes**

### CODE REVIEW

**Review by manasaling — APPROVED** [review]

File: tools/cli/factory.go [hunk 1]
```diff
@@ -139,6 +135,31 @@ func (b *clientFactory) ServerAdminClient(c *cli.Context) admin.Client {
 	return admin.NewThriftClient(serverAdmin.New(clientConfig))
 }
 
+// ServerFrontendClientForMigration builds a frontend client (based on server side thrift interface)
+func (b *clientFactory) ServerFrontendClientForMigration(c *cli.Context) frontend.Client {
```
Comment (Shaddoll) [inline tools/cli/factory.go:139]: the code is exactly the same as `ServerFrontendClient`, should `dispatcher` be replaced with something else?

File: tools/cli/factory.go [hunk 2]
```diff
@@ -139,6 +135,31 @@ func (b *clientFactory) ServerAdminClient(c *cli.Context) admin.Client {
...
+			apiv1.NewVisibilityAPIYARPCClient(clientConfig),
+		)
+	}
+	return frontend.NewThriftClient(serverFrontend.New(clientConfig))
+}
+
+// ServerAdminClientForMigration builds an admin client (based on server side thrift interface)
+func (b *clientFactory) ServerAdminClientForMigration(c *cli.Context) admin.Client {
+	b.ensureDispatcher(c)
+	clientConfig := b.dispatcher.ClientConfig(cadenceFrontendService)
```
Comment (Shaddoll) [inline tools/cli/factory.go:156]: same as above

File: tools/cli/factory.go [hunk 3]
```diff
@@ -70,35 +70,31 @@ type ClientFactory interface {
...
-	dispatcher      *yarpc.Dispatcher
-	logger          *zap.Logger
+	dispatcher          *yarpc.Dispatcher
+	dispatcherMigration *yarpc.Dispatcher
+	logger              *zap.Logger
 }
 
-// DEPRECATED don't use, only reserved for backward compatibility purposes
 // NewClientFactory creates a new ClientFactory
 func NewClientFactory() ClientFactory {
```
Comment (Shaddoll) [inline tools/cli/factory.go:90]: The method was marked as deprecated, is there a way to not use it?

File: tools/cli/factory.go [hunk 4]
```diff
@@ -159,19 +180,30 @@ func (b *clientFactory) ensureDispatcher(c *cli.Context) {
 	if b.dispatcher != nil {
 		return
 	}
+	b.dispatcher = b.newClientDispatcher(c, c.GlobalString(FlagAddress))
+}
+
+func (b *clientFactory) ensureDispatcherForMigration(c *cli.Context) {
```
Comment (Shaddoll) [inline tools/cli/factory.go:186]: this method is not used anywhere

File: tools/cli/factory.go (hunk 1 above)
Comment (shijiesheng) [reply tools/cli/factory.go:139]: good catch. fixed

File: tools/cli/factory.go (hunk 3 above)
Comment (shijiesheng) [reply tools/cli/factory.go:90]: it was deprecated in the previous CLI pr but then I realize it should not be deprecated

File: tools/cli/factory.go (hunk 4 above)
Comment (shijiesheng) [reply tools/cli/factory.go:186]: used now

**Review by Shaddoll — APPROVED** [review]

File: tools/cli/utils.go [hunk 5]
```diff
@@ -768,8 +768,8 @@ func newIndefiniteContext(c *cli.Context) (context.Context, context.CancelFunc)
 }
 
 func newTimedContext(c *cli.Context, timeout time.Duration) (context.Context, context.CancelFunc) {
-	if c.GlobalIsSet(FlagContextTimeout) {
```
Comment (shijiesheng) [inline tools/cli/utils.go:771]: this operation is not thread safe

**Pull request merged**